core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -838,9 +839,17 @@ public Object updateEvent() {
   }
 
   private void cleanUncommittedAppends(Set<ManifestFile> committed) {
-    if (cachedNewManifest != null && !committed.contains(cachedNewManifest)) {
-      deleteFile(cachedNewManifest.path());
-      this.cachedNewManifest = null;
+    if (cachedNewManifests != null) {
+      List<ManifestFile> cleanedNewManifests = Lists.newArrayList();
+      for (ManifestFile manifestFile : cachedNewManifests) {
+        if (!committed.contains(manifestFile)) {
+          deleteFile(manifestFile.path());
+        } else {
+          cleanedNewManifests.add(manifestFile);
+        }
+      }
+
+      this.cachedNewManifests = cleanedNewManifests;
```

Comment:
Sorry not sure if I follow the logic, why does cachedNewManifests need to be initialized to cleanedNewManifests?

Reply:
From the failed UTs, it seems it could call the `cleanUncommittedAppends` multiple times when retrying the commit. And we need to make sure the previous wroten manifest file is deleted.

Reply:
Added the comments.